python/torch_mlir/extras/onnx_importer.py

```diff
@@ -631,6 +758,218 @@ def tensor_proto_to_attr(self, tp: onnx.TensorProto) -> Attribute:
+            filter(f, op_schema.context_dependent_function_opset_versions),
+            default=None,
+        )
+        if ncd_function_version is None and cd_function_version is None:
+            # No relevant function definition
+            return None
+        elif ncd_function_version is not None and (
+            cd_function_version is None or cd_function_version < ncd_function_version
+        ):
+            specific_version = ncd_function_version
+            is_context_dependent = False
+        else:
+            specific_version = cd_function_version
+            is_context_dependent = True
+
+        # This is both a key for memoization of function importing and also a
+        # name mangling scheme, so it must include all information needed to
+        # uniquely identify a function and anything it might be parameterized
+        # over.
+        key = repr(
+            (
+                op_name,
+                op_domain,
+                opset_version,
+                input_type_protos,
+                # Though output types can be inferred from input types, it does
+                # not seem to be the case that there's only one legal set of
+                # outputs for a given set of inputs. When attemtping to always
+                # use onnx.shape_inference.infer_function_output_types instead
+                # of the caller-provided types, sometimes IR verification fails
+                output_type_protos,
+                # Avoid including the attributes twice (once on their own and
+                # once as part of the node) for context-dependent functions,
+                # avoid including unused parts of the node for other functions.
+                caller_node if is_context_dependent else caller_node.attribute,
+            )
+        )
+
+        existing = self._operator_function_map.get(key)
+        if existing is not None:
+            return existing
+
+        if is_context_dependent:
+            function_proto_str = (
+                op_schema.get_context_dependent_function_with_opset_version(
+                    specific_version,
+                    caller_node.SerializeToString(),
+                    [
+                        t.SerializeToString() if not isinstance(t, bytes) else t
+                        for t in input_type_protos
+                    ],
+                )
+            )
+        else:
+            function_proto_str = op_schema.get_function_with_opset_version(
+                specific_version
+            )
+        if not function_proto_str:
+            raise OnnxImportError(
+                f"Function lookup for {op_name}/{op_domain}/{specific_version}/{is_context_dependent} failed unexpectedly. This probably indicates a bug."
+            )
+        function_proto = onnx.onnx_pb.FunctionProto()
+        function_proto.ParseFromString(function_proto_str)
+
+        # An ONNX function may be polymorphic, parameterized over the types of
```

Comment:
I'd need to study the case further, but it feels like there should be a more unified way of doing this without an ONNX proto rewrite. But I get the issue and why this might be desired/obtuse to do the other way.

Perhaps more of a mechanical ask: Can you separate out the `convert_function_to_model_proto` vs inlining such a complicated transform like this?

Reply:
> I'd need to study the case further, but it feels like there should be a more unified way of doing this without an ONNX proto rewrite. But I get the issue and why this might be desired/obtuse to do the other way.

Are you imagining a way that doesn't involve creating a whole new model, or something else?

> Perhaps more of a mechanical ask: Can you separate out the convert_function_to_model_proto vs inlining such a complicated transform like this?

Ah, so then `get_operator_function` would only do the caching, schema lookup, and final conversion parts? I think I can do that.

Reply:
My gut feeling is that this is also somewhere the ONNX APIs could be improved.

- If shape inference could be run on a function and return the types of the interior nodes, not just the return types, we wouldn't have to turn the function into a model.
- If the Python API exposed shape inference on graphs (like the C++ does), not just on functions, we could skip the creating a model, albeit that step is trivial (it's making the graph that's a lot of work).
- I don't know if anyone else has needed to turn functions into graphs, but such a transform does feel like it belongs in ONNX itself if it has to go somewhere. It's easier for it to share code and prevent breakage.
- If the Python API exposed functions for attribute binding, we wouldn't have to reimplement it. See e.g. https://github.com/onnx/onnx/blob/88f8ef15cfaa3138d336f3502aed5018d802bf43/onnx/shape_inference/attribute_binder.h

Reply:
I outlined the transform to a new function, `_specialize_function_and_create_model`. That's definitely helped the readability of the code and let me give some things better name, thanks for that suggestion!